docs/architecture/design-proposals/dashboard/deployment-improvements.md

```diff
@@ -0,0 +1,197 @@
+4. Once the user clicks on "submit", we can use the `kubeapps` mapping to generate a JSON object. We can also use the JSON schema to validate the the processed object complies with the definition.
+
+## Upgrade Workflow
+
+The upgrade workflow will be similar to the deployment with some caveats:
+
+1. If the application to be upgraded doesn't contain a `values.schema.json`, behave as today. In case the current version contains the file, it should be included in new versions. If that's the case, go to the next step.
+2. If the schema contains the `kubeapps` key, use the current `values` to pre-populate the information of the form. If the schema doesn't contain that key, the renderization gets done by the automatic form generator so we may not be able to pre-fill the app information.
+3. Once the user selects a new version, the new `schema` gets loaded. For parameters existing in the previous version, re-fill the form with the previous values. In case there is a new parameter in the form, leave it empty (or use its default value). If a parameter is no longer used, ignore it.
+4. The logic to generate the final object and validate it is the same than for the deployment.
+
+## Development
+
+The amount of code/time required to implement this feature is considerably high. To avoid large PRs or temporary branches, we suggest to incrementaly implement this feature, merging changes on `master` but protecting the current status using a feature flag. If the feature flag is disabled (default behavior), the current deployment and upgrade form will be shown. For developers (or anyone that wants to test the feature), we will enable this flag to activate the new form. We can store this flag in the Kubeapps ConfigMap. Once the feature is finished, this flag will be removed from the configuration, enabling the feature for everyone.
+
+## Task definition
+
+This is the list of tasks that we foresee for this feature. This list may vary depending on new findings:
+
+- Create a values.schema.json for the desired applications.
+  - WordPress
+  - Apache
+  - PostgreSQL
+- Create a test step to ensure that if the `values.schema.json` exists, the default `values.yaml` complies with it.
+- Create the feature flag.
+- Allow to automatically convert the `values.schema.json` to a form (ignoring the `kubeapps` key).
+- Design and create the components for the rich view, using the `kubeapps` definition.
+- Design and create generic components in case `kubeapps` is defined with unknown parameters.
+- Adapt the DeploymentForm to render the `kubeapps` form.
+- Adapt the UpgradeForm to render the `kubeapps` form.
```

Comment:
YMMV, but as above, I'd include using `values.schema.json` for exactly what it's defined for first - validation. Then also storing the validated json in the redux state (so that it is updated by either the text component, or the form). Just a thought - see what you think.

Reply:
Let me add the validation as a standalone task.

Regarding the redux state: because of the reason I explained in the previous comment, I think it cannot be used for both the basic and the advanced deployment. In any case, both are in the same component (`DeploymentForm`) so it's not necessary to use redux, only using the `state` of the component will be enough.